build.rs

```diff
@@ -1,7 +1,52 @@
 use std::env;
 
+// This is not going to happen any more since we have a default feature definition in Cargo.toml
+//#[cfg(not(any(feature = "crypto_adaptor_openssl", feature = "crypto_adaptor_tongsuo")))]
+//compile_error! {
+//    r#"
+//    No cryptography adaptor is enabled!
+//
+//    In RustyVault, the real cryptographic operations are done via "crypto_adaptor"s.
+//
+//    A crypto adaptor is a module that conveys and translates high level cryptography
+//    operations like encryption, signing into the APIs provided by underlying cryptography
+//    libraries such as OpenSSL, Tongsuo and so forth.
+//
+//    At current stage, only one crypto_adaptor can be enabled at compilation phase and later
+//    be used at run-time. "crypto_adaptor"s are configured as 'feature's in the Cargo context.
+//
+//    Currently, the supported feature names of crypto adaptors are as follows, you can enable
+//    them by adding one '--features crypto_adaptor_name' option when running "cargo build":
+//        1. the OpenSSL adaptor: crypto_adaptor_openssl
+//        2. the Tongsuo adaptor: crypto_adaptor_tongsuo
+//    "#
+//}
+
+#[cfg(all(feature = "crypto_adaptor_openssl", feature = "crypto_adaptor_tongsuo"))]
+compile_error! {
+    r#"
+    Only one cryptography adapator can be enabled!
+
+    In RustyVault, the real cryptographic operations are done via "crypto_adaptor"s.
+
+    A crypto adaptor is a module that conveys and translates high level cryptography
+    operations like encryption, signing into the APIs provided by underlying cryptography
+    libraries such as OpenSSL, Tongsuo and so forth.
+
+    At current stage, only one crypto_adaptor can be enabled at compilation phase and later
+    be used at run-time. "crypto_adaptor"s are configured as 'feature's in the Cargo context.
+
+    Currently, the supported feature names of crypto adaptors are as follows, you can enable
+    them by adding one '--features crypto_adaptor_name' option when running "cargo build":
+        1. the OpenSSL adaptor: crypto_adaptor_openssl
+        2. the Tongsuo adaptor: crypto_adaptor_tongsuo
+    "#
+}
+
 fn main() {
     if let Ok(_) = env::var("DEP_OPENSSL_TONGSUO") {
         println!("cargo:rustc-cfg=tongsuo");
+    } else if cfg!(feature = "crypto_adaptor_tongsuo") {
+        println!("cargo:rustc-cfg=tongsuo");
```

Comment:
What you suggest is an implicit method to enable Tongsuo:

1. Build RustyVault with rust-openssl (patched to rust-tongsuo)
2. rust-tongsuo sets the `DEP_OPENSSL_TONGSUO` environment variable
3. build.rs notes the `DEP_OPENSSL_TONGSUO` variable and the print the 'cargo:rustc-cfg=tongsuo' stuff
4. the crypto Rust code use `#[cfg(tongsuo)]` and `#[cfg(not(tongsuo))]` to decide what to compile

This will cause chaos if there are other crypto adaptors in the future. If someday Tongsuo becomes very different and incomaptible with OpenSSL, things become difficult.

Reply:
Actually I think the previous check on `DEP_OPENSSL_TONGSUO` in not necessary anymore, since the user needs to explicitly specify `crypto_adaptor_tongsuo` when building RustyVault.

Reply:
Nah...if it is in this way, then `#[cfg(tongsuo)]` could also be changed to `#[cfg(feature = "crypto_adaptor_tongsuo")]` to avoid compilation warnings.

Reply:
As in the latest commit 44ec9a8, this is fixed by using feature name instead.